src/inference/src/dev/core_impl.hpp

```diff
@@ -0,0 +1,373 @@
+        config_["TARGET_FALLBACK"] = deviceName.substr(7);
+    } else if (deviceName_.find("MULTI:") == 0) {
+        deviceName_ = "MULTI";
+        config_[InferenceEngine::MultiDeviceConfigParams::KEY_MULTI_DEVICE_PRIORITIES] = deviceName.substr(6);
+    } else if (deviceName == "AUTO" || deviceName.find("AUTO:") == 0) {
+        deviceName_ = "AUTO";
+        if (deviceName.find("AUTO:") == 0) {
+            config_[InferenceEngine::MultiDeviceConfigParams::KEY_MULTI_DEVICE_PRIORITIES] =
+                deviceName.substr(std::string("AUTO:").size());
+        }
+    } else if (deviceName_.find("BATCH:") == 0) {
+        deviceName_ = "BATCH";
+        config_[CONFIG_KEY(AUTO_BATCH_DEVICE_CONFIG)] = deviceName.substr(6);
+    } else {
+        InferenceEngine::DeviceIDParser parser(deviceName_);
+        deviceName_ = parser.getDeviceName();
+        std::string deviceIDLocal = parser.getDeviceID();
+
+        if (!deviceIDLocal.empty()) {
+            config_[InferenceEngine::PluginConfigParams::KEY_DEVICE_ID] = deviceIDLocal;
+        }
+    }
+    return {deviceName_, config_};
+}
+
+class CoreImpl : public InferenceEngine::ICore, public std::enable_shared_from_this<InferenceEngine::ICore> {
+private:
+    mutable std::map<std::string, ov::InferencePlugin> plugins;
+    // Mutex is needed to prevent changes of dev mutexes map from different threads
+    mutable std::mutex global_mutex;
+    // Global mutex "" locks parallel access to pluginRegistry and plugins
+    // Plugin mutexes "plugin_name" lock access to code which changes configuration of particular plugin
+    mutable std::unordered_map<std::string, std::mutex> dev_mutexes;
+
+    std::mutex& get_mutex(const std::string& dev_name = "") const;
+    void add_mutex(const std::string& dev_name);
+
+    class CoreConfig final {
+    public:
+        struct CacheConfig {
+            std::string _cacheDir;
+            std::shared_ptr<InferenceEngine::ICacheManager> _cacheManager;
+        };
+
+        bool flag_allow_auto_batching = true;
+
+        void setAndUpdate(ov::AnyMap& config);
+
+        void setCacheForDevice(const std::string& dir, const std::string& name);
+
+        std::string get_cache_dir() const;
+
+        // Creating thread-safe copy of config including shared_ptr to ICacheManager
+        // Passing empty or not-existing name will return global cache config
+        CacheConfig getCacheConfigForDevice(const std::string& device_name,
+                                            bool deviceSupportsCacheDir,
+                                            std::map<std::string, std::string>& parsedConfig) const;
+
+        CacheConfig getCacheConfigForDevice(const std::string& device_name) const;
+
+    private:
+        static void fillConfig(CacheConfig& config, const std::string& dir);
+
+        mutable std::mutex _cacheConfigMutex;
+        CacheConfig _cacheConfig;
+        std::map<std::string, CacheConfig> _cacheConfigPerDevice;
+    };
```

Comment:
can we also move CoreConfig to a dedicated file?

Reply:
CoreConfig is a part of CoreImpl, it is why at the current moment I didn't move it to separate file